### PR TITLE
Create (stdlib) syntax for Coq stanza

### LINKED ISSUE #6163: Coq stanzas cannot be independent of `coq-stdlib`
This is a continuation of coq/coq#16510. Since Coq 8.17, Coq has been split into `coq-core` and `coq-stdlib`. It is now possible for Coq developments and plugins to only depend on `coq-core`. However, Dune does not properly support this yet.

To see this, have a Coq file with only `Declare ML Module "coq-core.plugins.ltac".` in it. This file does not depend on things in `coq-stdlib`  but it does depend on the Ltac library in `coq-core`. Then try to compile it with a stanza like this:

```
(coq.theory
 (name Test)
 (package test)
 (flags -noinit -boot)
)
```
This errors out with `cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package containing Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.` because `coqdep` is not passed the `-boot` flag. Adding `(boot)` does not help either: `Error: No rule found for theories/Init/Prelude.vo`.
It seems to me that
1. `-boot` should be passed to `coqdep` when it is passed as a flag.
2. When `(boot)` and `-noinit -boot` is specified, Dune should not go looking for `theories/Init/Prelude.vo`.
Alternatively, like @ejgallego suggested, a special `(stdlib no)` syntax could be introduced.

A concrete package that demonstrates this can be found here: https://github.com/coq-tactician/coq-tactician/tree/coqdev This development currently only compiles when full Coq is installed but errors out when only `coq-core` is installed.

To answer some questions of @ejgallego in coq/coq#16510:

> How does tactician work before the stdlib? Does the stdlib still needs to be compiled after it?

Tactician first compiles its own code that only depends on `coq-core`. Then, it can be injected into Opam's package installation process with `tactician inject` (which performs some clever tricks with `bwrap`). At that point `coq-stdlib` can be installed with support for Tactician using `opam install coq-stdlib`. Before the split, there was a special package `coq-tactician-stdlib` that would recompile and overwrite the stdlib because there was a chicken-and-the-egg problem between installing Tactician and Coq.

### PR DESCRIPTION
**Pull request by LasseBlaauwbroek**
Fixes #6163

@ejgallego could you make some comments?

### CODE REVIEW

File: src/dune_rules/coq_rules.ml
Comment (ejgallego) [inline src/dune_rules/coq_rules.ml:379]: Maybe this could go inside `coqc_file_flags` ?

**Comment by ejgallego** [comment]
This is def going in the good direction ! A test case is needed too.

It could be the case that the logic for this is better handled using the `bootstrap` logic, we need to modify that too, as when `(stdlib no)` is set, and a stdlib is in scope, we need to avoid adding it to the list of (implicit) lib deps.

File: src/dune_rules/coq_rules.ml
Comment (LasseBlaauwbroek) [reply src/dune_rules/coq_rules.ml:379]: Isn't it the case that `-noinit` should be included for `coqdep`? So there are separate code-paths for `coqdep` and `coqc`, which means that this cannot go into `coqc_file_flags`?

**Comment by LasseBlaauwbroek** [comment]
I tried testing this with Tactician while composed with Coq. But I didn't get very far, even without these new changes and without any kind of `-noinit` of `-boot` flag on Dune 3.4.1:
```
$ dune build user-contrib/Tactician/theories/Ltac1.vo
File "user-contrib/Tactician/theories/dune", line 1, characters 0-168:
1 | (coq.theory
2 |  (name Tactician)
3 |  (package coq-tactician)
4 |  (libraries
5 |    coq-tactician.record-plugin
6 |    coq-tactician.tactics-plugin
7 |    coq-tactician.ssreflect-plugin
8 |  )
9 | )
cannot guess a path for Coq libraries; please use -coqlib option or ensure you have installed the package containing Coq's stdlib (coq-stdlib in OPAM) If you intend to use Coq without a standard library, the -boot -noinit options must be used.
```
It seems that Dune is trying to build Tactician's library before the prelude is built. I do not remember this issue existing before...

**Comment by ejgallego** [comment]
@LasseBlaauwbroek oh, likely that's because Coq upstream doesn't ship `theories/dune` anymore, so Dune doesn't see Coq's stdlib. Can you try again and rename `theories/dune.disabled` in Coq's upstream?

Yes, the whole setup evolution of Coq + Dune is "interesting" to say the very least.

**Comment by LasseBlaauwbroek** [comment]
You mean rename `dune.disabled` to `dune`? If I do that, it does indeed compile... What is the story here? Is the disabled dune file expected to become default again?

**Comment by ejgallego** [comment]
> You mean rename `dune.disabled` to `dune`? If I do that, it does indeed compile... What is the story here? Is the disabled dune file expected to become default again?

Hopefully. The story is that Coq now generates `theories/dune` as to be able to fully migrate to Dune, as otherwise we were missing a few features such as configure-time native support and a couple other more.

The current state is interim only in order to be able to cleanup all the non-dune infrastructure in Coq. Also, we are blocked in Dune not allowing to include auto-generated dune files, otherwise we could have a much simpler setup with `theories/dune` including the auto-generated part.

**Comment by ejgallego** [comment]
> Error: in file Ltac1/Record.v, could not find META.coq-core.

This is a known issue, you need to use the hybrid `Declare ML Module` syntax still.

**Comment by LasseBlaauwbroek** [comment]
Here, you informed that I shouldn't use the hybrid syntax... https://github.com/coq/coq/issues/16510#issuecomment-1252299532

**Comment by LasseBlaauwbroek** [comment]
I'm also trying to write a test, but I cannot even get the existing tests to complete. I'm running `make test-coq`, it is unclear which Coq version I should use for the tests. All the ones that I've tried have failed the tests in different ways...

**Comment by ejgallego** [comment]
> Here, you informed that I shouldn't use the hybrid syntax... [coq/coq#16510 (comment)](https://github.com/coq/coq/issues/16510#issuecomment-1252299532)

Unfortunately that only works if you are not composing with the plugin in scope... If the plugin is not installed, you have this problem. Yeah the current state is not the best :/

Maybe you can keep that syntax for now, but issue a `dune build _build/install/default/lib/coq-core/META` first ?

There are two serious bugs for compositing with Coq >= 8.16 in current dune master, hopefully we can fix them asap.

**Comment by LasseBlaauwbroek** [comment]
Yes that works. But given that there are currently quite a few issues with composition, would it be reasonable to forego integrating this PR with the `bootstrap` logic? I cannot really test this, because it is unclear to me which issues are part of this PR and which are not...

If you can explain to me how to properly run the `make test-coq` tests, then I can add a small test for `(stdlib no)`, like you asked.

**Comment by ejgallego** [comment]
> If you can explain to me how to properly run the `make test-coq` tests, then I can add a small test for `(stdlib no)`, like you asked.

That should work out of the box [use the suggested opam switch in Dune's makefile]

> Yes that works. But given that there are currently quite a few issues with composition, would it be reasonable to forego integrating this PR with the bootstrap logic?

I think it is a good idea yes that `(stdlib no)` first works without a composed lib, but you'll need to hook into `boot_lib_flags` anways.

**Comment by LasseBlaauwbroek** [comment]
> That should work out of the box [use the suggested opam switch in Dune's makefile]

Thanks. I missed the `dev-switch` target.

> but you'll need to hook into boot_lib_flags anways.

I don't really know what that means. I don't really understand the full semantics of `(boot)` nor how you'd like me to modify it.

**Comment by ejgallego** [comment]
> I don't really know what that means. I don't really understand the full semantics of `(boot)` nor how you'd like me to modify it.

I mean that the place that you will have to tweak the flags generation is the function named `boot_lib_flags`. Once you tweak that function, composition should also work fine.

**Comment by LasseBlaauwbroek** [comment]
Another issue: It appears that dune is not passing the following flags to `coqc` in a composed build: ` -I plugins/syntax -I plugins/cc`.

**Comment by LasseBlaauwbroek** [comment]
> Once you tweak that function, composition should also work fine.

Unfortunately, I don't really know what this 'tweak' should be.

W.r.t. the composed build, I'm now stuck on this error:
```
File "./user-contrib/Tactician/theories/Ltac1/Record.v", line 5, characters 0-42:
Error:
System error: "/home/lasse/Documents/Projects/Tactician/coqdev/coq/_build/install/default/lib/coq-core/plugins/ltac/ltac_plugin.cmxs: No such file or directory"
```
It seems that plugins are not being properly promoted to the `install` directory?

**Comment by ejgallego** [comment]
Umm, that last one seems bizarre, but maybe you can try what I posted on the issue:

@LasseBlaauwbroek , Ali and I tried to fix the composition bug in https://github.com/ocaml/dune/pull/6165 , it was a bit tricky, but indeed, we added the flag to the buildable as we were going along. I think you can rebase your PR on top of that one, and add the tests, the syntax extension, and the documentation, and let us know how it works.

**Comment by ejgallego** [comment]
I'd suggest first adding a use case, use the one we added in the PR as a base if you want. Once the use case is working fine, you can see how it works on your own setup.

I'd be happy to try the tactician build if you point me out to the right tree.

**Comment by LasseBlaauwbroek** [comment]
Thanks @ejgallego, I'll try out the PR and see if it works.

With regard to writing tests and use cases, I think that we are talking past each other a little bit: I have absolutely no idea what I'm doing...
1. I do not understand what the intended semantics of `(boot)` is, neither do I know what 'the bug' related to this is. All I know is that `(boot)` is used to bootstrap Coq's standard library.
2. There seem to be so many bugs and issues that I'm have trouble seeing the trees in the forest.
3. I'm unsure what a test or use-case is supposed to test? Just that the `(stdlib no)` syntax exists and passes `-boot` to `-coqc` and `-coqdep`? Something related to `(boot)`? Something related to composition?

**Comment by LasseBlaauwbroek** [comment]
Here is the tree with some changes to Tactician that I'm currently using to try to make the composable build work: https://github.com/coq-tactician/coq-tactician/tree/no-stdlib-attempt

**Comment by ejgallego** [comment]
> 3\. I'm unsure what a test or use-case is supposed to test? Just that the `(stdlib no)` syntax exists and passes `-boot` to `coqc` and `coqdep`?

Yes. To merge this we need that test, a changes entry, and a documentation update.

> Something related to `(boot)`? Something related to composition?

That's optional IMHO, but easy to do if you want, start from some of the `compose-*boot*` tests, and for example verify that building a `(stdlib no)` library doesn't build the one that has `(boot)`.

By the way, the semantics of `(boot)` are not documented on purpose, but easy (once the PR fixing them is merged), it just tells dune that the library marked with boot has:
- a `$path/Init/` directory which is compiled with `-boot -noinit`
- all other libraries implicitly depend on it, and will implicitly load `Prelude.vo`

Actually both of these features could be modeled with smaller independent features:
- for the boot library, we could use per-file flags
- we could require the implicit dependency to be made explicit (we will soon)
- coqdep could add the missing `Prelude.vo` deps, however this is a bit harder as it is not backwards compatible.

**Comment by ejgallego** [comment]
> Here is the tree with some changes to Tactician that I'm currently using to try to make the composable build work: https://github.com/coq-tactician/coq-tactician/tree/no-stdlib-attempt

Your tree builds fine, but you need to still use the compat syntax, diff:
```diff
diff --git a/theories/Ltac1/Record.v b/theories/Ltac1/Record.v
index 0b1158c..1bc407c 100644
--- a/theories/Ltac1/Record.v
+++ b/theories/Ltac1/Record.v
@@ -2,11 +2,11 @@
 Definition private_constant_placeholder (x : Type) := x.
 Register private_constant_placeholder as tactician.private_constant_placeholder.
 
-Declare ML Module "coq-core.plugins.ltac".
-Declare ML Module "coq-core.plugins.firstorder".
-Declare ML Module "coq-core.plugins.extraction".
-Declare ML Module "coq-core.plugins.funind".
-Declare ML Module "coq-tactician.record-plugin".
+Declare ML Module "ltac_plugin:coq-core.plugins.ltac".
+Declare ML Module "firstorder_plugin:coq-core.plugins.firstorder".
+Declare ML Module "extraction_plugin:coq-core.plugins.extraction".
+Declare ML Module "funind_plugin:coq-core.plugins.funind".
+Declare ML Module "tactician_ltac1_record_plugin:coq-tactician.record-plugin".
 Export Set Default Proof Mode "Tactician Ltac1".
 
 Tactician Record Then Decompose.
diff --git a/theories/Ltac1/Tactics.v b/theories/Ltac1/Tactics.v
index 9461a3d..06e809c 100644
--- a/theories/Ltac1/Tactics.v
+++ b/theories/Ltac1/Tactics.v
@@ -1,6 +1,6 @@
 From Tactician Require Export Ltac1.Record.
 
-Declare ML Module "coq-tactician.tactics-plugin".
+Declare ML Module "tactician_ltac1_tactics_plugin:coq-tactician.tactics-plugin".
 
 Tactic Notation (at level 5) "search" "with" "cache" tactic3(t) := fail
   "'search with cache' has been renamed to 'synth with cache'".
```

**Comment by LasseBlaauwbroek** [comment]
Yes, that does indeed work (I feel stupid now for not thinking about that but the error messages were truly coming from everwhere). Thanks.

**Comment by LasseBlaauwbroek** [comment]
I tried to rebase this on top of #6165, but it looks like you've already integrated the new syntax introduced here?

**Comment by ejgallego** [comment]
> Yes, that does indeed work (I feel stupid now for not thinking about that but the error messages were truly coming from everwhere). Thanks.

Yes, it is very confusing to understand what is going on with the build. Incremental builds are harder than they look in the sense that it is very easy to have a case where things work intermittently, as in this case.

I'm gonna fix Dune so it doesn't need this awkward syntax.

> I tried to rebase this on top of https://github.com/ocaml/dune/pull/6165, but it looks like you've already integrated the new syntax introduced here?

Yup, I added a quick patch to be able to test tactician, but still we need to test and document the feature, as I proposed in the other PR let's make that PR about the fix (and adapting the internals for `use_stdlib`) and this PR about actually adding the field.

**Comment by LasseBlaauwbroek** [comment]
I added documentation, a simple test and a changelog item.

File: test/blackbox-tests/test-cases/coq/no-stdlib.t/foo.v
Comment (Alizter) [inline test/blackbox-tests/test-cases/coq/no-stdlib.t/foo.v:1]: Since this might fail for other reasons, could you remove the `Fail` and have the error message in the cram test. That way we know exactly which error happened.

File: test/blackbox-tests/test-cases/coq/no-stdlib.t/run.t
Comment (Alizter) [inline test/blackbox-tests/test-cases/coq/no-stdlib.t/run.t:1]: Could you also add a small explanation of the test above in a comment? (Anything you write without two spaces in a cram test is a comment).

File: test/blackbox-tests/test-cases/coq/no-stdlib.t/foo.v
Comment (Alizter) [reply test/blackbox-tests/test-cases/coq/no-stdlib.t/foo.v:1]: You could put the second fail in another .v file.

**Comment by Alizter** [comment]
The DCO job is complaining that you didn't signoff your commits which you can do with `git commit -s`.

Something like
```
git rebase -i upstream/main --exec 'git commit --amend --no-edit -n -s'
```

Might be useful. I typically have my git configured to do this automatically with
```
git config --global format.signoff true
```
(maybe you don't want it globally however)

File: doc/coq.rst
Comment (Alizter) [inline doc/coq.rst:106]: and coqdep right?

File: doc/coq.rst
Comment (Alizter) [inline doc/coq.rst:245]: Let's include the change from 0.5 here so that the doc doesn't look incomplete.

```suggestion
- ``0.5``: ``(libraries ...)`` field deprecated in favor of ``(plugins ...)`` field.
- ``0.6``: Support for ``(stdlib no)``.
```

File: src/dune_rules/coq_stanza.ml
Comment (Alizter) [inline src/dune_rules/coq_stanza.ml:67]:
```suggestion
        (Dune_lang.Syntax.since coq_syntax (0, 6) >>> enum [ ("yes", true); ("no", false) ])
```

File: doc/coq.rst
Comment (LasseBlaauwbroek) [reply doc/coq.rst:106]: Yeah, I thought that this was information that is kind of internal and does not need to be mentioned in the manual. Do you disagree?

Comment (Alizter) [reply doc/coq.rst:106]: OK let's keep it as is then.

**Review by Alizter — APPROVED** [review]

File: test/blackbox-tests/test-cases/coq/no-stdlib.t/run.t
Comment (ejgallego) [inline test/blackbox-tests/test-cases/coq/no-stdlib.t/run.t:4]: Maybe use `dune build --verbose` ? IMHO it is a good idea to guard here against silent flag changes.

Also let's add 2 tests where the boot library is in scope, one for yes and the other for now. You can just copy from the existing tests for boot copmosition.

**Review by ejgallego — COMMENTED** [review]
LGTM modulo maybe a couple of typos below? I think we should squash the commits too.

Also let's add a test where boot is in scope, and we use stdlib={yes,no}

File: test/blackbox-tests/test-cases/coq/no-stdlib.t/run.t
Comment (LasseBlaauwbroek) [reply test/blackbox-tests/test-cases/coq/no-stdlib.t/run.t:4]: I tried `--verbose` but I don't think that is going to work, because it's output is not deterministic:
```
-|  Running[0]: /usr/bin/ocamlc.opt -config > /tmp/build_01aa54_dune/dune_143239_output
+|  Running[0]: /usr/bin/ocamlc.opt -config > /tmp/build_784671_dune/dune_e90675_output
```

**Comment by LasseBlaauwbroek** [comment]
Squashed and fixed the comments you made.

File: test/blackbox-tests/test-cases/coq/no-stdlib.t/run.t
Comment (ejgallego) [reply test/blackbox-tests/test-cases/coq/no-stdlib.t/run.t:4]: Thanks! Maybe you can try -j1 and that will make the build deterministic? I'm unsure how this is handled in other setups, maybe instead of `--display=verbose` `--display=short` would suffice here?

Or we can just test the build of a single file, like `dune build --verbose foo.vo`, I think that should be actually enough.

Comment (Alizter) [reply test/blackbox-tests/test-cases/coq/no-stdlib.t/run.t:4]: Don't put `verbose` in tests it doesn't work.

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
@LasseBlaauwbroek Can you `./dune.exe build @fmt` to make the linter happy?

File: test/blackbox-tests/test-cases/coq/no-stdlib.t/run.t
Comment (LasseBlaauwbroek) [reply test/blackbox-tests/test-cases/coq/no-stdlib.t/run.t:4]: Yeah, I don't think that any of the verbose options are going to work. The issue is not so much with parallelism but with temporary directories.

**Comment by LasseBlaauwbroek** [comment]
Done.

File: test/blackbox-tests/test-cases/coq/no-stdlib.t/run.t
Comment (ejgallego) [reply test/blackbox-tests/test-cases/coq/no-stdlib.t/run.t:4]: Mmm, I see other tests are using `dune rules` to check this? But anyways this seems like a PITA so let's use `--display=short` instead for a single file, and check that nothing else is build when `(stdlib no)`

Comment (LasseBlaauwbroek) [reply test/blackbox-tests/test-cases/coq/no-stdlib.t/run.t]: Done.

**Comment by Alizter** [comment]
@LasseBlaauwbroek There is also @ejgallego's comment on the doc in my comment thread above.

**Comment by LasseBlaauwbroek** [comment]
I believe that I already addressed that one.

File: test/blackbox-tests/test-cases/coq/no-stdlib.t/run.t
Comment (ejgallego) [reply test/blackbox-tests/test-cases/coq/no-stdlib.t/run.t]: Thanks!

File: CHANGES.md
Comment (christinerose) [inline CHANGES.md:65]:
```suggestion
  plugin loading mechanism upstream (which now uses Findlib).
```

File: test/blackbox-tests/test-cases/coq/compose-boot-nostdlib.t/run.t
Comment (christinerose) [inline test/blackbox-tests/test-cases/coq/compose-boot-nostdlib.t/run.t:1]:
```suggestion
Testing composition of theories across a Dune workspace with a boot library and
```

File: test/blackbox-tests/test-cases/coq/compose-boot-nostdlib.t/run.t
Comment (christinerose) [inline test/blackbox-tests/test-cases/coq/compose-boot-nostdlib.t/run.t:2]:
```suggestion
importing ``stdlib`` enabled or disabled.
```

File: test/blackbox-tests/test-cases/coq/compose-boot-nostdlib.t/run.t
Comment (christinerose) [inline test/blackbox-tests/test-cases/coq/compose-boot-nostdlib.t/run.t:4]: Would it be okay to write "Composing library A..." instead of "a library A"? If so, please also change the following one for composing library B.

File: test/blackbox-tests/test-cases/coq/coqtop-root.t/run.t
Comment (christinerose) [inline test/blackbox-tests/test-cases/coq/coqtop-root.t/run.t:16]:
```suggestion
  Error: 'coq.theory' is available only when Coq is enabled in the ``dune-project``
```

File: test/blackbox-tests/test-cases/coq/coqtop-root.t/run.t
Comment (christinerose) [inline test/blackbox-tests/test-cases/coq/coqtop-root.t/run.t:17]:
```suggestion
  file. You must enable it using (using ``coq 0.6``) in your ``dune-project`` file.
```

File: test/blackbox-tests/test-cases/coq/github3624.t
Comment (christinerose) [inline test/blackbox-tests/test-cases/coq/github3624.t:18]:
```suggestion
  Error: 'coq.theory' is available only when Coq is enabled in the ``dune-project``
```

File: test/blackbox-tests/test-cases/coq/github3624.t
Comment (christinerose) [inline test/blackbox-tests/test-cases/coq/github3624.t:19]:
```suggestion
  file. You must enable it using (using ``coq 0.6``) in your ``dune-project`` file.
```

**Review by christinerose — APPROVED** [review]
suggestions for formatting, typo fixed

File: test/blackbox-tests/test-cases/coq/github3624.t
Comment (LasseBlaauwbroek) [reply test/blackbox-tests/test-cases/coq/github3624.t:18]: Autogenerated output.

File: test/blackbox-tests/test-cases/coq/github3624.t
Comment (LasseBlaauwbroek) [reply test/blackbox-tests/test-cases/coq/github3624.t:19]: Autogenerated output.

File: test/blackbox-tests/test-cases/coq/coqtop-root.t/run.t
Comment (Alizter) [reply test/blackbox-tests/test-cases/coq/coqtop-root.t/run.t:16]: But also technically correct since it is talking about the `coq` mode which is enabled in the `dune-project` file.

File: test/blackbox-tests/test-cases/coq/github3624.t
Comment (christinerose) [reply test/blackbox-tests/test-cases/coq/github3624.t:19]: Autogenerated from where? Like by `odoc`?

File: test/blackbox-tests/test-cases/coq/coqtop-root.t/run.t
Comment (Alizter) [reply test/blackbox-tests/test-cases/coq/coqtop-root.t/run.t:16]: Since it appears in a command line error message, making it monospace would waste space.

Comment (LasseBlaauwbroek) [reply test/blackbox-tests/test-cases/coq/coqtop-root.t/run.t:16]: @christinerose These are Cram (I/O) tests. The test system runs specified commands, and the output of the commands is tested against what is committed here. Hence, you cannot just change this output because then the tests would fail. If you would want to change this, you'd have to do it where the I/O is actually performed.

(As an additional note, I'd say that the required standard of copy-editing in tests is a bit lower than one would expect for, say, documentation.)

**Review by ejgallego — APPROVED** [review]
Great!

**Comment by LasseBlaauwbroek** [comment]
Thanks everyone!

File: test/blackbox-tests/test-cases/coq/coqtop-root.t/run.t
Comment (christinerose) [reply test/blackbox-tests/test-cases/coq/coqtop-root.t/run.t:16]: Understood. Please feel free to discard anything that's not useful. I was tagged to review, so I did my thing. 😄  .... Do you need me to go back and "approve" it before you can move forward?

File: test/blackbox-tests/test-cases/coq/github3624.t
Comment (Alizter) [reply test/blackbox-tests/test-cases/coq/github3624.t:19]: This is the command line error message of Dune.

File: test/blackbox-tests/test-cases/coq/coqtop-root.t/run.t
Comment (LasseBlaauwbroek) [reply test/blackbox-tests/test-cases/coq/coqtop-root.t/run.t:16]: No, the PR has already been merged and completed. Thanks for your comments @christinerose.

**Pull request merged**
